File: stacks/stacks.js
var Stack = function() {
  this.top = null;
};

var stackNode = function(data) {
  this.data = data;
  this.next = null;
};

Stack.prototype.push = function(data) {
  var t = new stackNode(data);
  t.next = this.top;
  this.top = t;
};

Stack.prototype.pop = function() {
  if (this.top === null) {
    console.log('Stack is empty!');
  }
  var item = this.top.data;
  this.top = this.top.next;
  return item;
}

Stack.prototype.peek = function() {
  if (this.top === null) {
    console.log('Stack is empty!')
  }
  return this.top.data;
};

Stack.prototype.isEmpty = function() {
  return this.top === null;
};

s = new Stack();

File: stacks/stacks.rb
class Stack
  def initialize
    @top = nil
  end

  def pop
    if @top == nil
      puts 'Stack is empty!'
    else
      item = @top
      @top = top.next
      item
    end
  end

  def push(data)
    n = Node.new(data)
    n.next = @top
    @top = n
  end

  def peek
    @top.data
  end

  def is_empty?
    @top == nil
  end
end

class Node
  attr_reader :data
  attr_accessor :next

  def initialize(data)
    @data = data
    @next = nil
  end
end

s = Stack.new
s.push('a')

puts s.peek